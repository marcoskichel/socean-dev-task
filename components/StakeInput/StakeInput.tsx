import {
  styled,
  TextField,
  InputLabel,
  Typography,
  Button,
} from '@mui/material';
import ArrowDownwardIcon from '@mui/icons-material/ArrowDownward';
import CurrencyIcon, { Currency } from '../CurrencyIcon/CurrencyIcon';

// Styles

const Wrapper = styled('div')(() => ({
  position: 'relative',
  width: '100%',
}));

const StyledTextField = styled(TextField)(() => ({
  'input::-webkit-outer-spin-button, input::-webkit-inner-spin-button': {
    '-webkit-appearance': 'none',
    margin: 0,
  },
  'input[type=number]': {
    '-moz-appearance': 'textfield',
    fontSize: '1.5rem',
    color: '#A7A7A7',
  },
  '.MuiInputBase-root': {
    backgroundColor: '#1C1C1C',
    height: '7rem',
    alignItems: 'end',
  },
  '.MuiInputLabel-formControl': {
    top: '-1.5rem',
  },
}));

const StyledInputLabel = styled(InputLabel)(({ theme }) => ({
  position: 'absolute',
  zIndex: 1,
  top: '1rem',
  left: '1rem',
  color: theme.palette.primary.main,
}));

const EndAdornmentWrapper = styled('div')(() => ({
  display: 'flex',
  flexDirection: 'column',
  height: '90%',
  justifyContent: 'space-around',
  paddingBottom: '0.5rem',
}));

const ArrowDownWrapper = styled('div')(() => ({
  position: 'absolute',
  left: 'calc(50% - 1.75rem)',
  bottom: '-1.75rem',
  width: '3.5rem',
  height: '3.5rem',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  backgroundColor: '#232323',
  borderRadius: '50%',
  border: '1px solid rgba(255, 255, 255, 0.23);',
  zIndex: 1,
}));

// Components

const ArrowDown = () => {
  return (
    <ArrowDownWrapper>
      <ArrowDownwardIcon />
    </ArrowDownWrapper>
  );
};

interface Props {
  value: string;
  currency: Currency;
  onChange: (value: string) => void;
}

const StakeInput = (props: Props) => {
  const { onChange, value } = props;

  /**
   * Avoid negative values
   * @param event The event object
   */
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.value.length > 0) {
      let value = Number(event.target.value);
      if (value < 0) {
        value = 0;
      }
      onChange(value.toString());
    } else {
      onChange('');
    }
  };

  const startAdornment = <CurrencyIcon currency={props.currency} />;

  const endAdornment = (
    <EndAdornmentWrapper>
      <Typography variant="caption" sx={{ whiteSpace: 'pre' }}>
        Balance: -
      </Typography>
      <Button
        variant="contained"
        size="small"
        disableElevation
        color="secondary"
        sx={{ background: 'rgba(255, 255, 255, 0.08)' }}
      >
        MAX
      </Button>
    </EndAdornmentWrapper>
  );

  return (
    <Wrapper>
      <StyledInputLabel>Amount</StyledInputLabel>
      <StyledTextField
        value={value}
        variant="outlined"
        type="number"
        onChange={handleChange}
        sx={{ width: '100%' }}
        InputProps={{
          startAdornment,
          endAdornment,
        }}
      />
      <ArrowDown />
    </Wrapper>
  );
};

export default StakeInput;
