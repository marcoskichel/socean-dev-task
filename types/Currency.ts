export enum Currency {
  SOL = 'SOL',
  scnSOL = 'scnSOL',
}
