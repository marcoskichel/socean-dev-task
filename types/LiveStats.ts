export interface LiveStats {
  title: string;
  value: string;
  hint: string;
  progress?: number;
}
