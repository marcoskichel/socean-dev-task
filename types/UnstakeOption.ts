export interface UnstakeOption {
  label: string;
  fee: number;
  value: number;
}
