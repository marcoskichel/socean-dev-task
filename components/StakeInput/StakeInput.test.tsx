import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Currency } from '../../types/Currency';
import StakeInput from './StakeInput';

describe('StakeInput', () => {
  it('renders', () => {
    const { asFragment } = render(
      <StakeInput currency={Currency.scnSOL} value="10" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('trigger onChange callback', () => {
    const onChange = jest.fn();
    render(<StakeInput currency={Currency.scnSOL} onChange={onChange} />);
    const input = screen.getByDisplayValue('0');
    userEvent.type(input, '1');

    expect(onChange).toHaveBeenLastCalledWith('1');
  });
});
