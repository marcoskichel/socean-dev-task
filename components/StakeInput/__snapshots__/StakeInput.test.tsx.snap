// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StakeInput renders 1`] = `
<DocumentFragment>
  <div
    class="css-1jke4yk"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary css-esg5jd-MuiFormLabel-root-MuiInputLabel-root"
    >
      Amount
    </label>
    <div
      class="MuiFormControl-root MuiTextField-root css-1ybss1w-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-2ehmn7-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          class="css-1wowqjn"
        >
          <span
            style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 32px; height: 32px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
          >
            <img
              alt="Currency"
              data-nimg="fixed"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
        <input
          aria-invalid="false"
          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-152mnda-MuiInputBase-input-MuiOutlinedInput-input"
          id="mui-1"
          type="number"
          value="10"
        />
        <div
          class="css-1dqf7fm"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-4akox3-MuiTypography-root"
          >
            Balance: -
          </span>
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-disableElevation MuiButtonBase-root  css-1aeyr1q-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            MAX
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-nnbavb"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="css-vvfmul"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="ArrowDownwardIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m20 12-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
